### PR TITLE
Add OpenAI-compatible HTTP API (/v1/chat/completions, /v1/models)  

### PR DESCRIPTION
**Pull request by jaswinder6991**
## Summary
                                                                                                                       
  - Adds OpenAI-compatible /v1/chat/completions and /v1/models endpoints to the web gateway, allowing any standard     
  OpenAI client library to use IronClaw as an LLM backend by simply changing the base_url
  - Implements both non-streaming and streaming (SSE) response modes with full tool/function calling support
  - Returns OpenAI-format errors ({ error: { message, type, param, code } }) with proper HTTP status codes (401, 429,
  400, 503)
  - Adds docker-compose.yml for local PostgreSQL with pgvector

## Test plan

  - 14 unit tests covering type conversions, serialization, and edge cases (cargo test openai_compat)
  - 9 integration tests with mock LLM provider testing full HTTP path (auth, streaming, tools, error cases)
  - Manual verification against live NEAR AI API (DeepSeek-V3.1)
  - Verified drop-in compatibility with Python openai library
  - Verify no regressions on existing gateway endpoints

### CODE REVIEW

File: src/channels/web/openai_compat.rs [hunk 1]
```diff
@@ -0,0 +1,1011 @@
...
+use super::server::GatewayState;
+
+// ---------------------------------------------------------------------------
+// OpenAI request types
+// ---------------------------------------------------------------------------
+
+#[derive(Debug, Deserialize)]
+pub struct OpenAiChatRequest {
+    #[allow(dead_code)]
+    pub model: String,
```
Comment (serrrfirat) [inline src/channels/web/openai_compat.rs:33]: **High: `model` field is silently ignored**

`#[allow(dead_code)]` means the client-specified model is completely discarded — the handler always uses `llm.active_model_name()` (lines 422, 530, 766). This breaks the OpenAI contract where clients choose the model.

Two options:
1. **Forward it**: pass `req.model` into `CompletionRequest` / `ToolCompletionRequest` so the provider can honor it (or call `set_model()` before the call)
2. **Reject mismatches**: if the provider only supports one model, return a 404 `model_not_found` when the requested model doesn't match `active_model_name()`

Silently ignoring it is the worst option — a client sending `model: "gpt-4"` will get back a response claiming `model: "claude-3-5-sonnet"` with no warning.

File: src/channels/web/openai_compat.rs [hunk 2]
```diff
@@ -0,0 +1,1011 @@
...
+        Ok(Json(response).into_response())
+    }
+}
+
+/// Handle streaming responses.
+///
+/// The current `LlmProvider` returns complete responses (no streaming method).
+/// We call the LLM in a spawned task and simulate chunked delivery by splitting
+/// the response into word-boundary chunks. True token streaming can be added
+/// later by extending `LlmProvider` with a `complete_stream()` method.
```
Comment (serrrfirat) [inline src/channels/web/openai_compat.rs]: **High: Streaming is simulated — clients get no latency benefit**

The handler calls `llm.complete()` / `llm.complete_with_tools()` which both block until the full response is ready, then splits the result into ~20 char word-boundary chunks (line 684). From the client's perspective, time-to-first-token equals time-to-full-response — the SSE chunking is cosmetic.

This is a correctness problem for any client using streaming for progressive rendering (most of them). Two suggestions:
1. **Short-term**: document this limitation prominently in the API response headers or in a `x-ironclaw-streaming: simulated` header, so clients know what they're getting
2. **Medium-term**: add `complete_stream()` to `LlmProvider` that returns a `Stream<Item=...>` and wire it through here

The code comment acknowledges this, but external clients won't read the source.

File: src/channels/web/openai_compat.rs [hunk 3]
```diff
@@ -0,0 +1,1011 @@
...
+                    let err_data = serde_json::to_string(&OpenAiErrorResponse {
+                        error: OpenAiErrorDetail {
+                            message: e.to_string(),
+                            error_type: "server_error".to_string(),
+                            param: None,
+                            code: None,
+                        },
+                    })
+                    .unwrap_or_default();
+                    let _ = tx.send(Ok(Event::default().data(err_data))).await;
```
Comment (serrrfirat) [inline src/channels/web/openai_compat.rs]: **Medium: Streaming errors are indistinguishable from content**

When the LLM call fails mid-stream, the error is sent as a regular SSE `data:` event containing a JSON error object (lines 619-630). The HTTP status is already committed as 200 at this point.

OpenAI SDK clients don't expect error JSON in the data stream — they'll either crash trying to parse it as a chunk or silently swallow it. The OpenAI convention is to send an error event with `"error"` in the `data` field or use a distinct SSE event type.

Consider:
```
Event::default().event("error").data(err_data)
```
so clients can distinguish error events from content chunks.

**Review by serrrfirat — DISMISSED** [review]

**Review by serrrfirat — CHANGES_REQUESTED** [review]
## Review — PR #31

**Files**: 7 files, 1578+/0- | **Assessment**: REQUEST_CHANGES

Well-structured OpenAI-compatible API with good test coverage (mock provider, streaming, tool calls, error cases). The simulated streaming via word-boundary chunking is a reasonable interim approach.

### Findings

| Severity | Finding |
|----------|---------|
| P2 | No request size limit on `/v1/chat/completions` — a malicious client could send a massive messages array. The gateway's existing 1MB body limit may cover this but verify the new routes go through it |
| P2 | No rate limiting on `/v1` endpoints — the gateway's chat endpoints have rate limiting but these new routes don't |
| P2 | `model` field is accepted but silently ignored — all requests route to whatever LLM provider is configured. Document this behavior or implement model routing |
| P3 | `docker-compose.yml` has hardcoded `POSTGRES_PASSWORD: ironclaw` — add a comment noting this is dev-only |
| P3 | Simulated streaming is documented in code but clients may notice the chunking pattern. Consider a doc note |

### What's good
- Comprehensive integration tests with mock LLM provider
- Proper OpenAI error format mapping from LlmError variants
- Tool call support (both streaming and non-streaming)
- `[DONE]` sentinel correctly sent for streaming

**Comment by jaswinder6991** [comment]
The web gateway currently has no rate limiting on any endpoint (the HTTP webhook channel does, but the gateway    doesn't). Should I add rate limiting across all gateway routes, not just the /v1 ones?                               
                                                            
Also, model mismatch now returns 404 model_not_found. A good follow up for per-request model routing tracked in #49.
